### PR TITLE
Add `be` prefix & make some functions more generic

### LINKED ISSUE #153: Add `be` (Big Endian) prefix to the relevant conversion functions in the `utils` package

### LINKED ISSUE #174: Buffers returned by `leBigintToBuffer` and `beBigintToBuffer` should have dynamic length
**Is your feature request related to a problem? Please describe.**

The `leBigintToBuffer` and `beBigintToBuffer` functions only return buffers with 32 bytes. `leBigintToBuffer` was originally used inside the `eddsa-poseidon` package, where it only needed to return a buffer with a static size (32 bytes). Since those functions are part of the `utils` now, they should be generic and support different options. This is a breaking change as it will be mandatory to define the bytes if one needs to convert a number to a buffer with static length.

**Describe the solution you'd like**

These functions should add an additional optional parameter: `length`. If `length` is not defined or if it is less than the number of bytes of the `bigint` to be converted, the minimum number of bytes required for the conversion will be used.  

Example:
```ts
export function beBigintToBuffer(n: bigint, length?: number): Buffer {
    const hex = bigIntToHexadecimal(n)

    // Calculate the minimum buffer length required to represent 'n' in bytes.
    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
    const minLength = Math.ceil(hex.length / 2)

    // Use the provided length or the calculated minimum length, whichever is greater.
    const bufferLength = length ? Math.max(length, minLength) : minLength

    // Allocate buffer of the desired size, filled with zeros.
    const buffer = Buffer.alloc(bufferLength, 0)

    const fromHex = Buffer.from(hex, "hex")
    fromHex.copy(buffer, bufferLength - fromHex.length)

    return buffer
}
```

### PR DESCRIPTION
**Pull request by cedoor**
## Description

This PR mainly covers the functions of the `conversions` module of the `@zk-kit/utils` library. It essentially does the following:

1. Adds the prefix `be` to functions that do conversions with hexadecimals or buffers in big-endian.
2. Renames the `Bigint`/`bigint` occurrences in function names to `BigInt`/`bigInt`.
3. Adds a `size` parameter to the `leBigintToBuffer` and `beBigintToBuffer` functions to define the number of bytes of the buffer returned by those function.
4. Adds tests for the module's `conversions` functions.
5. Documents all the code in the `conversions` module.

This PR introduces breaking changes! See the related issues for more info.

## Related Issue(s)

Closes #153
Closes #174

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] I have made corresponding changes to the documentation
-   [x] My changes generate no new warnings
-   [x] I have run `yarn prettier` and `yarn lint` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/utils/src/conversions.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ * It also includes variations for both big-endian (`be`) and little-endian
+ * (`le`) conversions. It is important to note that when there is no prefix,
+ * the order of bytes is always big-endian.
+ */
+
+import { BigNumberish } from "./types"
+import { isHexadecimal, isStringifiedBigint } from "./number-checks"
+
+/**
+ * Converts a bigint to a hexadecimal string ensuring even length.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:19]: If the function name makes byte-order explicit, the documentation should say what the byte order is, in order to describe what "be" stands for.  E.g. "Converts a bigint to a hexadecimal string ensuring even length.  Uses big-endian byte order."

This same comment applies to all of the functions in this file, so I'm not going to repeat it.  They should all make byte order explicit in the documentaiton.

File: packages/utils/src/conversions.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+    return hex
+}
+
+/**
+ * Converts a bigint to a hexadecimal string. Alias for beBigIntToHexadecimal.
+ * @param n The bigint value to convert.
+ * @returns The hexadecimal representation of the bigint.
+ */
+export function bigIntToHexadecimal(n: bigint): string {
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:39]: For hexadecimal stringification of a number, I actually don't think byte order needs to be explicit.  It doesn't hurt, but I think in this case there actually is an obvious default.  Numbers represented as strings always put the most significant digit first.  Importantly these functions are really outputting digits (4 bits at a time) not bytes.

Relatedly, in this case it's weird to make BE byte order explicit, but not provide any LE alternatives.

File: packages/utils/src/conversions.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+/**
+ * Converts a bigint to a hexadecimal string ensuring even length.
+ * @param n The bigint value to convert.
+ * @returns The hexadecimal representation of the bigint.
+ */
+export function beBigIntToHexadecimal(n: bigint): string {
+    let hex = n.toString(16)
+
+    // Ensure even length.
+    if (hex.length % 2 !== 0) {
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:27]: I don't know all the use cases, but I wonder if what they really need isn't padding to an even number, but padding out to a specific size (e.g. 32 bytes) which should be passed as an argument.  This doesn't need to be resolved in this PR, it's just something which occurs to me as an easy misuse.

File: packages/utils/src/conversions.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+ * Converts a bigint to a hexadecimal string. Alias for beBigIntToHexadecimal.
+ * @param n The bigint value to convert.
+ * @returns The hexadecimal representation of the bigint.
+ */
+export function bigIntToHexadecimal(n: bigint): string {
+    return beBigIntToHexadecimal(n)
+}
+
+/**
+ * Converts a hexadecimal string to a bigint, ensuring the hex string starts with '0x'.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:44]: "ensuring the hex string starts with 0x" isn't behavior which is visible to the caller, so it doesn't seem like the right thing to call out in documentation.

Alternative suggestions: "Coverts a hexadecimal string to a bigint.  The input is interpreted as hexadecimal with or without a `0x` prefix."

File: packages/utils/src/conversions.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+ * Converts a hexadecimal string to a bigint. Alias for beHexadecimalToBigint.
+ * @param hex The hexadecimal string to convert.
+ * @returns The bigint representation of the hexadecimal string.
+ */
+export function hexadecimalToBigint(hex: string): bigint {
+    return beHexadecimalToBigint(hex)
+}
+
+/**
+ * Converts a buffer to a bigint by interpreting the buffer as a hexadecimal string.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:65]: "by interpreting the buffer as a hexadecimal string" is an implementation detail which I think is actually confusing here.  It suggests to me that as caller I should be passing an array of strings.  The buffer isn't being intereted as a hex string, it's being converted to a hex string internally via the call to toString() below.

Alternative suggestion: "Converts a buffer of bytes to a bigint using big-endian byte order."

File: packages/utils/src/conversions.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * Converts a buffer to a bigint. Alias for beBufferToBigint.
+ * @param b - The buffer to convert.
+ * @returns The bigint representation of the buffer's contents.
+ */
+export function bufferToBigint(b: Buffer): bigint {
+    return beBufferToBigint(b)
+}
+
+/**
+ * Converts a bigint to a buffer, assuming big-endian byte order, and fills with zeros if necessary.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:83]: I'd say "using" rather than "assuming" big-endian byte order.

"assuming" to me would suggest that big-endian is a property of the bigint, while it's actually a property of the buffer introduced by the conversion.

Same comment applies to the other uses of "assuming" in this file, so I won't repeat it.

File: packages/utils/src/conversions.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+ * @returns The buffer representation of the bigint.
+ */
+export function beBigintToBuffer(n: bigint, length?: number): Buffer {
+    const hex = bigIntToHexadecimal(n)
+
+    // Calculate the minimum buffer length required to represent 'n' in bytes.
+    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
+    const minLength = Math.ceil(hex.length / 2)
+
+    // Use the provided length or the calculated minimum length, whichever is greater.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:95]: This behavior needs to be documented or it'll be surprising.  Perhaps the argument should be called minLenth?
Alternatively if minLength > length this function could throw an exception.  Incorrect buffer lengths can be a source of security bugs, so I think checking and failing loudly might be the best approach.  (The current approach is already much safer than truncating, though.)

File: packages/utils/src/conversions.ts [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    return beBigintToBuffer(n)
+}
+
+/**
+ * Converts a BigNumberish type to a buffer, handling various input types and converting them to bigint first if necessary.
+ * @param n The BigNumberish value to convert.
+ * @returns The buffer representation of the BigNumberish value.
+ */
+export function beBigNumberishToBuffer(n: BigNumberish): Buffer {
+    if (
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:122]: The logic here could be delegated to `bigNumberishToBigint` to avoid duplication. That would simplify this function to only need one special case for input which is already a buffer.

  if (typeof n === "Buffer") {
    return n as Buffer;
  }
  return bigintToBuffer(bignumberishToBigint(n));

File: packages/baby-jubjub/src/baby-jubjub.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { F1Field, bigintToHexadecimal, bufferToBigint, leBigintToBuffer, leBufferToBigint, scalar } from "@zk-kit/utils"
+import { F1Field, bigIntToHexadecimal, bufferToBigint, leBigintToBuffer, leBufferToBigint, scalar } from "@zk-kit/utils"
```
Comment (artwyman) [inline packages/baby-jubjub/src/baby-jubjub.ts:1]: This name change seems inconsistent with other functions in this library.  E.g. leBigintToBuffer doesn't capitalize the `I` in Bigint, so why should this one function do so?

It would be good to standardize on one rule for whether that I should be capitalized.

File: packages/utils/src/conversions.ts [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+export function bigintToBuffer(n: bigint): Buffer {
+    return beBigintToBuffer(n)
+}
+
+/**
+ * Converts a BigNumberish type to a buffer, handling various input types and converting them to bigint first if necessary.
+ * @param n The BigNumberish value to convert.
+ * @returns The buffer representation of the BigNumberish value.
+ */
+export function beBigNumberishToBuffer(n: BigNumberish): Buffer {
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:121]: These "bignumberish" helpers are another case where it's odd not to have LE alternatives.  Unlike the hex digits above, I do think this is a case where it's good to make the byte order explicit, since the use of a buffer means there is a choice.

File: packages/utils/src/conversions.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+ * @returns The buffer representation of the bigint in little-endian.
+ */
+export function leBigintToBuffer(n: bigint, length?: number): Buffer {
+    const hex = bigIntToHexadecimal(n)
+
+    // Calculate the minimum buffer length required to represent 'n' in bytes.
+    // Each hexadecimal character represents 4 bits, so 2 characters are 1 byte.
+    const minLength = Math.ceil(hex.length / 2)
+
+    // Use the provided length or the calculated minimum length, whichever is greater.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:192]: Similar to above, this behavior should be documented, and I have a similar question as to whether it's better to rename length -> minLength, or to throw on overflow.

File: packages/utils/tests/conversions.test.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        it("Should convert a bigint to a BE hexadecimal", async () => {
+            const result = hexadecimalToBigint(testHex1BE)
+
+            expect(result).toBe(testBigInt1BE)
+        })
+    })
+
+    describe("# bigNumberishToBuffer", () => {
+        it("Should convert a big numberish to a BE buffer", async () => {
+            const result = bigNumberishToBuffer(testBigInt1BE)
```
Comment (artwyman) [inline packages/utils/tests/conversions.test.ts:40]: Should have tests exercising the other possible input types.

File: packages/utils/tests/conversions.test.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+            expect(result).toStrictEqual(Buffer.from(testBytes1.slice(1)))
+        })
+    })
+
+    describe("# bigNumberishToBigint", () => {
+        it("Should convert a BE big numberish to a bigint", async () => {
+            const result = bigNumberishToBigint(Buffer.from(testBytes1))
```
Comment (artwyman) [inline packages/utils/tests/conversions.test.ts:48]: Again, testing of the other input types would be welcome.

File: packages/utils/tests/conversions.test.ts [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+            expect(result).toBe(testBigInt1BE)
+        })
+    })
+
+    describe("Bigint to/from Buffer Conversions", () => {
```
Comment (artwyman) [inline packages/utils/tests/conversions.test.ts:54]: Consider adding tests to exercise the overflow behavior I commented on elsewhere (i.e. what happens when the bigint is too big for the specified buffer size).  Your code-coverage analysis will probably point that out as an untested branch.

**Review by artwyman — CHANGES_REQUESTED** [review]
Looks good overall.  A lot of comments primary around documentation and testing.

File: packages/utils/src/conversions.ts (hunk 1 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:19]: Right! I've added an explanation in the module description but it would be better to be explicit in those functions too 👍🏽

File: packages/utils/src/conversions.ts (hunk 3 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:27]: From what I know `Buffer` needs it. `Buffer.from("02", "hex")` is not the same as `Buffer.from("2", "hex")`.

It shouldn't make any difference in the other use-cases.

Feel free to open another issue for that feature if you want 👍🏽

Comment (cedoor) [reply packages/utils/src/conversions.ts:27]: BTW there's an issue for replacing `Buffer` with `Uint8Array`, to make these libraries fully-compatible with browsers -> https://github.com/privacy-scaling-explorations/zk-kit/issues/144.

Your input would be really appreciated. I'll work on that asap.

File: packages/utils/src/conversions.ts (hunk 2 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:39]: I see. I'd tend to prefer leaving it without any prefix then. If we realize someone needs this conversion with little-endian byte order we could make it more explicit. What do you think?

File: packages/utils/src/conversions.ts (hunk 4 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:44]: Right, definitely clearer! Thanks

File: packages/utils/src/conversions.ts (hunk 8 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:122]: Good catch! The following code should be fine:
```ts
 if (n instanceof Buffer) {
        return n
 }

return bigintToBuffer(bigNumberishToBigint(n))
```

I'd like to add more checks on types here. But it'll be part of another PR: https://github.com/privacy-scaling-explorations/zk-kit/pull/178 (feel free to start taking a look).

File: packages/baby-jubjub/src/baby-jubjub.ts (hunk 9 above)
Comment (cedoor) [reply packages/baby-jubjub/src/baby-jubjub.ts:1]: Right! IMO we should use "BigInt" everywhere, as it's the same name JS uses.

Comment (cedoor) [reply packages/baby-jubjub/src/baby-jubjub.ts]: I've renamed them. I started marking breaking changes in my commits: https://github.com/privacy-scaling-explorations/zk-kit/pull/175/commits/6858f66f6778bb9656ca97a391f8128ff354d0bf.

File: packages/utils/src/conversions.ts (hunk 10 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts]: Makes sense! I've added `leBigNumberishToBuffer`, but not sure about `beBigNumberishToBigInt`.

File: packages/utils/src/conversions.ts (hunk 2 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts:40]: I agree.

File: packages/utils/src/conversions.ts (hunk 10 above)
Comment (artwyman) [reply packages/utils/src/conversions.ts]: Why not?  I feel like there should be symmetry.

File: packages/utils/src/conversions.ts [hunk 15]
```diff
@@ -5,7 +5,7 @@
  * conversions involving bigints, hexadecimals and buffers.
  * The module is structured with clear function naming to indicate
  * the conversion direction (e.g., `bigIntToHexadecimal` for BigInt
- * to hexadecimal, `bufferToBigint` for buffer to bigint) and employs
+ * to hexadecimal, `bufferToBigInt` for buffer to bigint) and employs
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:8]: I somehow missed this module documentation on my first review.  I think I'm used to skipping over big comment blocks at the tops of files which are usually copyright text.

Regardless, I like this description - very clear.

File: packages/utils/src/conversions.ts [hunk 16]
```diff
@@ -41,111 +42,128 @@ export function bigIntToHexadecimal(n: bigint): string {
...
+    return beBufferToBigInt(b)
 }
 
 /**
- * Converts a bigint to a buffer, assuming big-endian byte order, and fills with zeros if necessary.
+ * Converts a bigint to a buffer and fills with zeros if necessary.
+ * It uses big-endian byte order.
  * @param n The bigint to convert.
- * @param length The number of bytes of the buffer to return.
+ * @param size The number of bytes of the buffer to return.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:97]: You should document what happens if this parameter is absent.  Also document the error thrown if the size is too small.
I like the new version of the argument and logic, thanks.

File: packages/utils/src/conversions.ts [hunk 17]
```diff
@@ -41,111 +42,128 @@ export function bigIntToHexadecimal(n: bigint): string {
...
 }
 
 /**
- * Converts a bigint to a buffer. Alias for beBigintToBuffer.
- * @param n - The bigint to convert.
- * @returns The buffer representation of the bigint.
+ * Converts a bigint to a buffer and fills with zeros if necessary.
+ * It uses little-endian byte order.
+ * @param n The bigint to convert.
+ * @param size The number of bytes of the buffer to return.
```
Comment (artwyman) [inline packages/utils/src/conversions.ts:126]: Same need for documentation here of absent size, and error behavior.

**Review by artwyman — APPROVED** [review]
Looks good, thanks.  A few more suggestions in comments, but they're minor, so not going to block you on another round of review from me across timezones.

**Review by ctrlc03 — APPROVED** [review]

File: packages/utils/src/conversions.ts (hunk 15 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts:8]: I also skit it usually. It will be more visible on the TypeDoc website I guess.

File: packages/utils/src/conversions.ts (hunk 10 above)
Comment (cedoor) [reply packages/utils/src/conversions.ts]: Right, a bignumber-ish could also be a buffer indeed.

Comment (artwyman) [reply packages/utils/src/conversions.ts]: The case where the BigNumberish is itself a buffer is a tricky one here.  It raises the question of whether the BE/LE prefix applies to the input buffer, or the output buffer, or both.  I think it's too complicated to introduce 4 separate options to separately specify input and output byte order.  So I could see two approaches which could be consistent here:

1) The LE/BE specifier applies to **both** input and output.  That means when a conversion is needed, use the selected byte order.  But if the input is already a buffer, it's assumed to already be in the desired byte order, so it can be passed through unchanged.
2) The LE/BE specifier applies to the "Buffer" output.  BigNumberish is assumed (and should be documented) to always assume a BE buffer.  In this case, if the input is a Buffer you'd need to reverse it as you say.

I think #1 is the simplest option for this function.  It's really an argument justifying how it already works.  I think #2 might be more consistent broadly, because all the functions which accept a BigNumberish are doing BE conversion by default.

Realizing this complexity makes me feel more like it's best to not provide a LE version after all.  I'd just leave the function as-is, and document what it does for clarity.  In particular, document that if the input is a buffer it's left unmodified, and if the input is not a buffer it's converted using BE order.  So long as callers know that, they can figure out what they need to do.

Comment (cedoor) [reply packages/utils/src/conversions.ts]: > Realizing this complexity makes me feel more like it's best to not provide a LE version after all. I'd just leave the function as-is, and document what it does for clarity.

I agree. I removed the LE version and added more documentation to explain how they work.
